// zcatalyst-sdk-node/src/utils/constants.js
'use strict';

const envOverride = require('./helper').envOverride;

module.exports = Object.freeze({
	ACCOUNTS_ORIGIN: envOverride('CATALYST_ACCOUNTS_URL', 'https://accounts.zoho.com'),
	CATALYST_ORIGIN: envOverride('CATALYST_CONSOLE_URL', 'https://catalyst.zoho.com'),
	PRODUCT_NAME: 'baas',
	API_VERSION: 'v1',
	CATALYST_AUTH_ENV_KEY: 'CATALYST_AUTH',
	CREDENTIAL_SUFFIX: 'catalyst/application_auth.json',
	AUTH_HEADER: 'Authorization',
	PROJECT_KEY_NAME: 'PROJECT_ID',
	DEFAULT_APP_NAME: '[DEFAULT]',
	CATALYST_CONFIG_ENV_KEY: 'CATALYST_CONFIG',
	CREDENTIAL_TYPE: {
		token: 'token',
		ticket: 'ticket'
	},
	CREDENTIAL_HEADER: {
		admin_type: 'x-zc-admin-cred-type',
		user_type: 'x-zc-user-cred-type',
		admin_token: 'x-zc-admin-cred-token',
		user_token: 'x-zc-user-cred-token'
	},
	CREDENTIAL_USER: {
		admin: 'admin',
		user: 'user'
	},
	PROJECT_HEADER: {
		id: 'x-zc-projectid',
		domain: 'x-zc-project-domain',
		key: 'x-zc-project-key'
	},
	INIT_TYPE: {
		applogic: 'applogic',
		basicio: 'basicio'
	},
	CLIENT_ID: 'client_id',
	CLIENT_SECRET: 'client_secret',
	EXPIRES_IN: 'expires_in',
	AUTH_URL: 'auth_url',
	REFRESH_URL: 'refresh_url',
	REDIRECT_URL: 'redirect_url',
	REFRESH_TOKEN: 'refresh_token',
	ACCESS_TOKEN: 'access_token',
	CONNECTOR_NAME: 'connector_name',
	GRANT_TYPE: 'grant_type',
	CODE: 'code'
});
